accepted/cranelift-assembler.md

```diff
@@ -0,0 +1,368 @@
+   the work of encoding, register-allocating, pretty-printing, etc., is being done elsewhere, not in
+   `cranelift-codegen`.
+2. Create mapping functions from `cranelift-codegen`'s current types (e.g., `GprMemImm`) to the
+   assembler's more restrictive types (e.g., `AssemblerGpr` or `AssemblerImm8`). These mappings are
+   likely to involve extra conversion work that may temporarily slow down compilation; this proposal
+   expects later analysis and optimization to fix this. (E.g., adding the `External` variant bumps
+   up the `MInst` size from 40 bytes to 48 bytes which should have _some_ instruction cache
+   effects).
+3. Over time, replace lowering rules in ISLE that use the current logic `(rule (lower ...))` to use
+   the new `External` instructions. As this happens, we will stumble over the various
+   inconsistencies in the current "category" paradigm; e.g., some `NInst` variants are actually
+   _sequences_ of instructions. We anticipate this piece-by-piece transformation may require new
+   abstractions, e.g., `ExternalSequence`.
+4. Eventually, all lowering of CLIF to machine instructions will use `cranelift-assembler-*` at
+   which point we can remove the `External` variant and any assembler logic from
+   `cranelift-codegen`.
+
+This high-level plan should replace the "category" instructions with concrete machine lowerings, but
+it does not foresee _all_ issues along the way. The prototype does, however, already highlight two
+potential issues that we raise here for discussion sake:
+
+- __abstraction leaks__: to lower from CLIF &mdash; a high-level SSA IR &mdash; to low-level machine
+  code we truly must "lower" from one abstraction level to another. E.g., @cfallin and @abrown have
+  already identified that `cranelift-codegen`'s decision to add a non-existent but necessary output
+  operand to x64 instructions causes problems for the current `cranelift-assembler-*` prototype
+  (though we're working on this). This proposal would argue that, for sanity's sake, we should avoid
+  leakage of this kind of `codegen`-necessary logic into the assembler, if at all possible.
+- __verbosity__: the assembler also materializes all the possible machine instructions, something
+  that `cranelift-codegen` did not do; one benefit of `cranelift-codegen` in this sense is that it
+  hid some of this detail behind the `MInst` "categories." As the assembler brings this to light,
+  one possibility is that the lowering ISLE becomes very verbose (i.e., one line per machine
+  instruction totals roughly 20 lines for CLIF's `band`). @abrown's predilection is to surface this,
+  @cfallin would prefer to hide it. One approach is to generate extra ISLE helpers that group
+  together common patterns; the `*.isle` files would be available somewhere but not in tree. Another
+  option would be to extend ISLE with macro-like functionality. Feedback is welcome on this point!
```

Comment:
I think this is also touching on the tension/impedance mismatch between "assembler library that directly reflects `x86_64`" and "final machine-specific backend IR for Cranelift". At the ISLE lowering rules level, I don't want to think about formats at all. I want to think about whole instructions and that's it. At the register operand collection and encoding levels, I don't care so much about whole instructions, and formats can be great shortcuts for deduplicating code.

I think we should prioritize and optimize for ISLE lowering rules. The latter things will ideally be automatically generated from the DSL describing the instructions, and so the duplication aspect doesn't affect our maintenance burden (although it admittedly could affect Cranelift's run time performance, the time it takes to build Cranelift, and the size of a Cranelift binary; nonetheless I'm not too concerned in this case).

Reply:
I'm not sure I completely understand what you mean: let's talk about "categories." `x64_and` is a category of instructions, not an x64 instruction. `x64_andb`, despite being more specific, is also a category of instructions. `x64_andb_mi` is an actual instruction. I interpret your comment about formats to actually echo something @cfallin and I have talked about, which is that actually in ISLE we _do_ want to use categories, e.g., `x64_and`. Is that what you mean?